src/SDK/SmartDetectorsShared/Clients/AnalysisServicesFactory.cs

```diff
@@ -184,22 +186,69 @@ private async Task<IReadOnlyList<string>> GetResourceIdsForLogAnalyticsAsync(IRe
+                    // If the resource is a KubernetesService cluster then get its associated workspace directly. Otherwise all workspaces in its subscription must be retrieved.
+                    if (resource.ResourceType == ResourceType.KubernetesService)
                     {
-                        subscriptionWorkspaces = await this.azureResourceManagerClient.GetAllResourcesInSubscriptionAsync(subscriptionId, new[] { ResourceType.LogAnalytics }, cancellationToken);
-                        this.subscriptionIdToWorkspaces[subscriptionId] = subscriptionWorkspaces;
+                        // Retrieve the specific workspace the target Kubernetes cluster belongs to.
+                        if (!this.aksClusterIdToWorkspaces.TryGetValue(resource.ToString(), out ResourceIdentifier workspace))
+                        {
+                            // Try to get the workspaces from the cache, and if it isn't there, use the Azure Resource Manager client
+                            ResourceProperties resourceProperties = await this.azureResourceManagerClient.GetResourcePropertiesAsync(resource, cancellationToken);
+
+                            // (get values from Json in a case-insensitive way)
+                            if (((resourceProperties.Properties?.GetValue("addonProfiles", StringComparison.CurrentCultureIgnoreCase) as
+                                JObject)?.GetValue("omsagent", StringComparison.CurrentCultureIgnoreCase) as
+                                JObject)?.GetValue("enabled", StringComparison.CurrentCultureIgnoreCase)?.ToObject<bool>() ?? false)
+                            {
+                                string idstring = (((resourceProperties.Properties.GetValue("addonProfiles", StringComparison.CurrentCultureIgnoreCase) as
+                                    JObject)?.GetValue("omsagent", StringComparison.CurrentCultureIgnoreCase) as
+                                    JObject)?.GetValue("config", StringComparison.CurrentCultureIgnoreCase) as
+                                    JObject)?.GetValue("loganalyticsworkspaceresourceid", StringComparison.CurrentCultureIgnoreCase).ToString();
+
+                                // idstring will only be null if ARM reported that the omsagent was enabled but there was no logAnalyticsWorkspaceResourceID.
+                                // Throw an exception if this is the case, the OMS agent is probably misconfigured.
+                                if (idstring != null)
+                                {
+                                    workspace = ResourceIdentifier.CreateFromResourceId(idstring);
+                                    this.aksClusterIdToWorkspaces[resource.ToString()] = workspace;
+                                }
+                                else
+                                {
+                                    throw new TelemetryDataClientCreationException("Specified cluster does not have monitoring add-on enabled");

```

Comment:
>Specified cluster does not have monitoring add-on enabled [](start = 84, length = 57)

More specific error message please - for example: "The specified cluster is misconfigured"